### PR TITLE
Support parsing Oracle EXPLAIN PLAN sql #27145

### LINKED ISSUE #27145: Support parsing Oracle EXPLAIN PLAN sql
# Background
Hi community,
This issue is for #26878.

ShardingSphere parser engine helps users parse a SQL to get the AST (Abstract Syntax Tree) and visit this tree to get SQLStatement (Java Object). Currently, we are planning to enhance the support for Oracle SQL parsing in ShardingSphere.

More details:
https://shardingsphere.apache.org/document/current/en/reference/sharding/parse/

# Task
This issue is to support more oracle sql parse, as follows:
```sql
EXPLAIN PLAN FOR
SELECT country, prod, year, sales
FROM sales_view
WHERE country IN ('Italy', 'Japan')
MODEL UNIQUE DIMENSION
  PARTITION BY (country) DIMENSION BY (prod, year) MEASURES (sale sales)
  IGNORE NAV RULES UPSERT AUTOMATIC ORDER
  (sales['All Products', 2003] = 200000,
   sales['Y Box', 2003] = sales['Bounce', 2003] * 0.25,
   sales['Bounce', 2003] = sales['Y Box', 2003] + 
    (sales['Bounce', 2002] / SUM(sales)[ANY, 2002] * 2 * 
     sales['All Products', 2003]));
```

```sql
EXPLAIN PLAN FOR
SELECT country, prod, year, sales
FROM sales_view
WHERE country IN ('Italy', 'Japan')
MODEL UNIQUE DIMENSION
  PARTITION BY (country) DIMENSION BY (prod, year) MEASURES (sale sales)
  RULES UPSERT AUTOMATIC ORDER
  (sales['Y Box', 2003] = sales['Bounce', 2003] * 0.25,
   sales['Bounce', 2003] = sales['Bounce', 2002] / SUM(sales)[ANY, 2002] * 2 *
   sales['All Products', 2003],
   sales['All Products', 2003] = 200000);
```

```sql
EXPLAIN PLAN FOR
SELECT p.ad_textdocs_ntab
FROM print_media_part p;
```

```sql
EXPLAIN PLAN FOR
SELECT e.employee_id, j.job_title, e.salary, d.department_name
    FROM employees e, jobs j, departments d
    WHERE  e.employee_id < 103
       AND e.job_id = j.job_id 
       AND e.department_id = d.department_id;
```

```sql
EXPLAIN PLAN FOR 
  SELECT * FROM emp_comp 
  WHERE hire_date = TO_DATE('15-FEB-1998', 'DD-MON-YYYY');
```

# Process
1. First confirm that this is a correct oracle sql syntax, if not please ignore;
2. Compare SQL definitions in Oficial SQL Doc and ShardingSphere SQL Doc;
3. If there is any difference in ShardingSphere SQL Doc, please correct them by referring to the Official SQL Doc;
4. Run mvn install the current_file_module;
5. Check whether there are any exceptions. If indeed, please fix them. (Especially xxxVisitor.class);
6. Add new corresponding SQL case in SQL Cases and expected parsed result in Expected Statment XML;
7. Run SQLParserParameterizedTest to make sure no exceptions.

# Relevant Skills
1. Master JAVA language
2. Have a basic understanding of Antlr `g4` file
3. Be familiar with Oracle SQLs

### PR DESCRIPTION
**Pull request by ilyasahsan123**
Fixes #27145.

Changes proposed in this pull request:
  - Support parsing Oracle EXPLAIN PLAN sql #27145

---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

### CODE REVIEW

**Comment by ilyasahsan123** [comment]
hi, @strongduanmu @FlyingZC, Could you please take a look at this? Thanks!

**Comment by strongduanmu** [comment]
> hi, @strongduanmu @FlyingZC, Could you please take a look at this? Thanks!

Thank you for your contribution, I will review this pr.

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/DMLStatement.g4 [hunk 1]
```diff
@@ -686,7 +686,8 @@ subquery
 modelExpr
     : (numberLiterals ASTERISK_)? ((measureColumn LBT_ (condition | expr) (COMMA_ (condition | expr))* RBT_) 
     | (aggregationFunction LBT_ (((condition | expr) (COMMA_ (condition | expr))*) | (singleColumnForLoop (COMMA_ singleColumnForLoop)*) | multiColumnForLoop) RBT_) 
-    | analyticFunction) (PLUS_ modelExpr | ASTERISK_ numberLiterals (ASTERISK_ modelExpr)?)?
+    | analyticFunction) ((PLUS_|SLASH_) LP_? modelExpr* RP_? | ASTERISK_ numberLiterals (ASTERISK_ modelExpr)?)?
+    | expr
```
Comment (strongduanmu) [inline parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/DMLStatement.g4:690]: Can you add sql parser test case for this rule change?

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/DMLStatement.g4 [hunk 2]
```diff
@@ -686,7 +686,8 @@ subquery
 modelExpr
     : (numberLiterals ASTERISK_)? ((measureColumn LBT_ (condition | expr) (COMMA_ (condition | expr))* RBT_) 
     | (aggregationFunction LBT_ (((condition | expr) (COMMA_ (condition | expr))*) | (singleColumnForLoop (COMMA_ singleColumnForLoop)*) | multiColumnForLoop) RBT_) 
-    | analyticFunction) (PLUS_ modelExpr | ASTERISK_ numberLiterals (ASTERISK_ modelExpr)?)?
+    | analyticFunction) ((PLUS_|SLASH_) LP_? modelExpr* RP_? | ASTERISK_ numberLiterals (ASTERISK_ modelExpr)?)?
```
Comment (strongduanmu) [inline parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/DMLStatement.g4:689]: Besides, can you add one space before and after `|` symbol.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/DMLStatement.g4 (hunk 1 above)
Comment (ilyasahsan123) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/DMLStatement.g4:690]: As previously mentioned, the test case has already been added. You can find it in the following link: 
- https://github.com/apache/shardingsphere/blob/master/test/it/parser/src/main/resources/sql/supported/dal/explain.xml#L74-L87

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/DMLStatement.g4 (hunk 2 above)
Comment (ilyasahsan123) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/DMLStatement.g4]: Sure thing, I have made the adjustment.

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/DMLStatement.g4 (hunk 1 above)
Comment (strongduanmu) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/DMLStatement.g4:690]: Thanks for remind, it's ok now.

**Review by strongduanmu — APPROVED** [review]
Looks great, merged. @ilyasahsan123 Looking forward to your next PR.

**Comment by ilyasahsan123** [comment]
Thanks, @strongduanmu for have a look at this. Sure thing, I'll sent another pull request.

**Pull request merged**